### PR TITLE
ghcWithPackages: add withLLVM

### PR DESCRIPTION
**Pull request by abbradar**
Following https://github.com/NixOS/nixpkgs/issues/6707 . cc @peti

### CODE REVIEW

**Comment by peti** [comment]
I'm worried about the fact that the appropriate LLVM version for each `ghc` is hidden somewhere inside the argument list of the `ghc` expression. We normally have that kind of version-specific know-how in `pkgs/development/haskell-modules/configuration-ghc-x.yz.nix`. It's confusing, IMHO, that LLVM stuff lives in an entirely different place.

**Comment by abbradar** [comment]
Good point! How about this version, then? It also cleans up several overrides  Note that this sets particular LLVM version corresponding to compiler for all package set, but I think that it's the right thing to do.

File: pkgs/development/haskell-modules/with-packages-wrapper.nix [hunk 1]
```diff
@@ -25,15 +29,20 @@ assert stdenv.lib.versionOlder "6.12" ghc.version;
...
-  hasLibraries  = stdenv.lib.any (x: x.isHaskellLibrary) paths;
+  paths         = filter (x: x ? isHaskellLibrary) (closePropagation packages);
+  hasLibraries  = any (x: x.isHaskellLibrary) paths;
+  # CLang is needed on Darwin for -fllvm to work:
+  # https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/code-generators.html
+  llvm          = makeSearchPath "bin"
+                  ([ llvmPackages.llvm ]
+                   ++ optional stdenv.isDarwin llvmPackages.clang);
 in
-if paths == [] then ghc else
```
Comment (peti) [inline pkgs/development/haskell-modules/with-packages-wrapper.nix]: This if-else clause prevents the wrapper from failing in case someone instantiates `ghcWithPackages` with an empty package, or with a package set that is non-empty but doesn't contain any actual Haskell libraries, i.e.:

  ghcWithPackages (hp: [hp.cabal-install])

Do these cases still work correctly with the LLVM patches applied?

Comment (abbradar) [inline pkgs/development/haskell-modules/with-packages-wrapper.nix]: Hm, `ghc` and `ghci` both work properly for me a simple "Hello World" and this derivation:
```
(pkgs.haskellngPackages.ghcWithPackages (pkgs: [ pkgs.cabal-install ])).override {
  withLLVM = true;
}
```
Actually I don't see why should there be problems; I thought this was an optimization.

File: pkgs/development/haskell-modules/with-packages-wrapper.nix [hunk 2]
```diff
@@ -1,7 +1,11 @@
-{ stdenv, ghc, packages, buildEnv, makeWrapper, ignoreCollisions ? false }:
+{ stdenv, ghc, llvmPackages, packages, buildEnv
+, makeWrapper
+, ignoreCollisions ? false, withLLVM ? false }:
+
+with stdenv.lib;
```
Comment (peti) [inline pkgs/development/haskell-modules/with-packages-wrapper.nix]: IMHO, this change is unrelated and I'd rather not merge it (and the other changes that depend on it) as part of the LLVM updates. I realize it's a matter of taste, but I'd rather not have symbols in scope if I don't have to; I tend to avoid  using`with`.

**Comment by peti** [comment]
Test builds are now running at http://hydra.cryp.to/jobset/nixpkgs/haskell-updates starting with [evaluation 292319](http://hydra.cryp.to/eval/292319). I've merged your changes from 3e6f271 into my local repository for those builds. I've also added 4028898, f56acc0, and 3d94719 to activate the changes you've made.

**Comment by abbradar** [comment]
Nice, thanks! I haven't known that no `with` is your policy -- if you want, I can make a patch which reverts things to "before `with`" state.

**Comment by peti** [comment]
Merged to `master`.
